src/uu/cp/src/cp.rs

```diff
@@ -898,9 +939,32 @@ fn copy_directory(root: &Path, target: &Target, options: &Options) -> CopyResult
     let mut hard_links: Vec<(String, u64)> = vec![];
 
     for path in WalkDir::new(root) {
-        let path = or_continue!(or_continue!(path).path().canonicalize());
+        let p = or_continue!(path);
+        let is_symlink = fs::symlink_metadata(p.path())?.file_type().is_symlink();
+        let path = if options.no_dereference && is_symlink {
+            // we are dealing with a symlink. Don't follow it
+            env::current_dir().unwrap().join(normalize_path(p.path()))
```

Comment:
Please don’t use `unwrap()`.  This function returns `CopyResult`, so you can just add another error to that if you need.